### DISCUSSION TITLE
Implementation / Dependecies of WiFi.setHostname()

### DISCUSSION THREAD

**Discussion by Sral931** [earlephilhower/arduino-pico#1637] [Q&A] [answered]
I'm having an issue with my Raspberry Pico W. I don't have the code with me right now, but I can upload it later if required.

The issue I'm having is as such: I can't get my Pico W to show up under it's hostname in Station mode, which is of course dependent on the Pico's and my Router's DNS handling.

I'm basically doing the following:

```
#include <WiFi.h>
#include <WebServer.h>
/* other includes like OneWire library, Pico/stdlib, etc */

const char* SSID = "MY_SSID";
const char* PASSWORD = "MY_PASSWORD";
const char* HOSTNAME = "PicoMug";

/* other code to init pins and constants on core0 here */

setup1(){
    // Init Serial port and wait for max 10s
    Serial.begin(115200);
    while (!Serial && millis() < 10000){
        digitalWrite(LED_BUILTIN, HIGH);
        delay(100);
        digitalWrite(LED_BUILTIN, LOW);
        delay(800);
    }

    // init Wifi
    WiFi.setHostname(HOSTNAME);
    WiFi.begin(SSID, PASSWORD);
    
    while (WiFi.connected() != WL_CONNECTED) {
        Serial.println("Connecting Wifi ...");
        digitalWrite(LED_BUILTIN, HIGH);
        delay(100);
        digitalWrite(LED_BUILTIN, LOW);
        delay(100);
        digitalWrite(LED_BUILTIN, HIGH);
        delay(100);
        digitalWrite(LED_BUILTIN, LOW);
        delay(700);
    }
    Serial.print("IP-address: ");
    Serial.println(WiFi.localIP());

    /* other code to init a Webserver */

    // send message to core0 to start normal operation
    rp2040.fifo.push(1)
}
```

My problem is that my I can't reach the Pico under that hostname, I always have to enter it's IP in the browser or for pings. The code works without crashing, so any errors in the code above is from my faulty memory. The Data aquisition on Core0 runs perfectly, as does the Webpage serving on Core1.

Am I doing something wrong with the handling of the WiFi-library, or do I have to use another mDNS library to actively request a DNS entry on the DNS-server like another user did on #1257  ?

Appreciate any help with this !

My project is already crowded, so I don't realy want to use a display to constantly display the IP on the outside or connect a cable to read the IP through the Serial port. If nothing works I'll probably have to send the last IP number as binary code through the Builtin_LED ^^
![Tasse](https://github.com/earlephilhower/arduino-pico/assets/105347332/091dd357-1c80-4515-abd3-35a990ff2333)

**Comment by earlephilhower** [comment]
The Pico does not handle real DNS lookups, that's up to your router configuration or whatever your  DNS server settings are set to (i.e. sometimes it's set to Google's `8.8.8.8` or Cloudflare `1.1.1.1` ).

Assuming you've set your router to be the DNS server for your network (and any DHCP replies it sends out) then you might also need to append whatever domain name you've set up as your local net in it.

MDNS basically lets every machine on the network serve its own hostname to anyone that asks.  So, that might be your best (only) bet if your DNS server is set to a global internet one like above, or if your router sometimes doesn't register DHCP names with DNS.  It uses some (small) CPU but because it's multicast should be reachable on your net under the `myname.local` address.

**Reply by Sral931** [reply]
Thanks a lot !

I'm not learned at all in these things, but what I do know is that my Raspberry Pi Zero with it's ReefPi installation is reachable under ReefPi.local . Not sure though it that's a similar case of the Raspberry Pi or ReefPi doing mDNS though, or my Router doing it.

Do you know of a library that works well with the Arduino-Pico core ?
I've looked at the [DDNS_Generic library](https://www.arduinolibraries.info/libraries/ddns_generic) and think that's more or less what I need, since I'll be mostly using the mug at work, where we have a dedicated DNS server.

**Reply by earlephilhower** [reply, marked as answer]
Yes, `xxx.local` is MDNS and not port 53 DNS.

MDNS is already included in the core, see the examples in https://github.com/earlephilhower/arduino-pico/tree/master/libraries/LEAmDNS

**Reply by Sral931** [reply]
Again, thanks a lot ! Works beautifully at home using the MDNS code from https://github.com/earlephilhower/arduino-pico/discussions/1257

Let's see how it does at work in the Guest-WiFi with the companies Firewall and DNS-Server possibly interfering.
